src/main/kotlin/org/rust/ide/intentions/ContractModule.kt

```diff
@@ -0,0 +1,32 @@
+package org.rust.ide.intentions
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiFile
+import com.intellij.psi.impl.file.PsiFileImplUtil
+import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesUtil
+import org.rust.lang.core.psi.impl.RustFileImpl
+import org.rust.lang.core.psi.util.RustModules
+
+class ContractModule : IntentionAction {
+
+    override fun getFamilyName()        = "Contract module structure"
+    override fun getText()              = "Contract Module"
+    override fun startInWriteAction()   = true
+
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?) =
+        file is RustFileImpl
+            && file.name == RustModules.MOD_RS
+            && file.containingDirectory?.children?.size == 1
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        file ?: return
+        var dst = file.parent?.parent
+        var parent = file.parent
+        val fileName = parent?.name + ".rs"
+        PsiFileImplUtil.setName(file, fileName)
+        MoveFilesOrDirectoriesUtil.doMoveFile(file, dst)
```

Comment:
>It looks like PsiDirectory.getName() is @NotNull.

Yeah, but the parent itself can be `null`.

>I decorated the var expressions with ?: return, although I have no idea if that's idiomatic Kotlin.

For binding nullable variables, you can use `?.let` or ` ?: retrun` patterns. I generally prefer `?.let` for simple cases and returns for anything more complex. The remind me of `>>=` and `do` from Haskell.

Here, I would use something like 

```Kotlin
val parent = file.parent ?: return
val grandpa = parent.parent ?: return
```